### PR TITLE
[NetworkManager] move NetworkManager-specific content to a new plugin.

### PR DESCRIPTION
**Pull request by robbmanes**
As NetworkManager is being further developed as a stand-alone service,
it is likely that new output will be necessary.  Therefore, further
cluttering the existing networking plugin may be inadvisable and the
two should be separated.

This commit simply removes any NetworkManager-specific output from the
networking plugin and moves it to it's own NetworkManager plugin.  From
there expansion on to the service can be done easily without fear of
disruption of the underlying networking plugin.

Signed-off-by: Robb Manes rmanes@redhat.com

### CODE REVIEW

**Comment by robbmanes** [comment]
This seems, to me at least, like a good idea to partition off nmcli content from the networking plugin and include it in its own portion.  This reduces the size/complexity of the networking plugin (particularly when new features like namespaces get introduced) and then provides an easier way to deal with the faster-moving NetworkManager development.

I haven't added anything new; I have plans, depending on how they pan out, to perhaps make the NetworkManager content collect more per-connection information (for troubleshooting daemon issues by UUID, for example, as I've had to do) but we capture a lot of that already and I didn't want to overcomplicate anything.  Still, a separate plugin may be a good idea.

Test sosreport is available at:

http://people.redhat.com/rmanes/sos-patch/sosreport-rmanes.networkmanager01-20150914195445.tar.xz

I've tested on a number of systems just to ensure both the original networking plugin is fully intact and no errors (that I can see) exist in the NetworkManager plugin exist.

If you don't agree or feel this is unnecessary, I completely understand and yield to your decision.

**Comment by battlemidget** [comment]
I'm not sure this makes sense. Can you expand on what you mean by `As NetworkManager is being further developed as a stand-alone service,`. Hasn't NetworkManager always been a standalone service?

**Comment by robbmanes** [comment]
Ah yes, sorry for vagueness.  I mean since the latest additions to NetworkManager, particularly adding bonding support, bridge support, teaming and VLAN support now all being intermixed since v1 releases, the complexity in the issues we're seeing with NetworkManager are getting more specialized/complex and it may make sense to have them handled by a different plugin that can be enabled/disabled at will, particularly when NetworkManager is not present.

I was (poorly) trying to elaborate on how NetworkManager is now growing from something that was universally recommended to be turned off due to lack of features or reliability in favor of the initscripts of a particular distribution into a more "legitimate" service after v1.  More effort, insofar as I can see from working with it, is going into making NetworkManager accessible and the default mechanism in which networking is configured, so individual troubleshooting of the service apart from the underlying network infrastructure accessed via `ip` may be beneficial.  In this way, we could inspect NetworkManager `connections` and `devices` independent of the networking plugin, which (to my mind, and please feel free to set me straight) should provide all of the information pertaining to the underlying networking, but not necessarily data pertaining to the service that manages and/or assigns them in the same manner you would not expect floating-ip data to be collected via ipvs commands by anything other than LVS and/or keepalived plugins.

Also, if you feel this is unnecessary, particularly now and/or depending on your release plans, I understand, as (I at least) do not terribly often get NetworkManager related issues requiring much more data collection than what is already in an sosreport, but it does happen from time to time and I figured separating the two may be a good idea for future expansion.  Concerning priority, definitely bottom-of-the-barrel at best if you decide this is worthwhile.  Thanks!

TL;DR: I fear NetworkManager or the networking plugin may grow in complexity in the future, and was attempting to remove it from the networking plugin so that the option to not collect nmcli output is present, and perhaps make it a bit easier to work on them independently from each other.

**Comment by bmr-cymru** [comment]
I tend to agree with Rob (tbh I didn't like adding the nmcli stuff to networking in the first place but it was done under release pressure).

My main concern, and one reason I've not tried to do this already, is that I feat they may end up wanting to share data or state and that we don't have the infrastructure to do that effectively at the moment.

I'll take a look at the proposals though & comment here.

**Comment by robbmanes** [comment]
In its current incarnation, I don't think they have much to share, as NetworkManager uses its own structures (`connection` and `device`) for manipulation of underlying devices independent of what would need to be gathered by, say, `ip`.  I couldn't think of a use case that they'd need to be combined, nor do I see one being mentioned in the upstream sections of NetworkManager.  That being said, in future editions, who knows.

**Comment by bmr-cymru** [comment]
No: it isn't really a current concern more that as we try to add the higher-level representations of the data we collect I half expect that we'll want to share at least some abstractions (if not actual data / implementation) between these - e.g. a notion of IP addresses etc.

Anyway, that's a bridge we can cross when we come to it. I think the current work is looking like a good improvement on its own.

**Comment by robbmanes** [comment]
@bmr-cymru Oh, I wasn't disagreeing; I meant that for the immediate future of the NetworkManager project, insofar as I know from working in it a bit, the output and/or structure doesn't change too radically.  I understand that the collection we do here can be more reliant (requiring cross-plugin commands, etc) in the future and has to be taken into account, likely in ways I can't possibly think of.

**Comment by battlemidget** [comment]
ack

**Comment by battlemidget** [comment]
conflicts with master

**Comment by bmr-cymru** [comment]
This fell off our radar a while back trying to get a release finalised. @robbmanes where do you think we stand with this today? There's been a fair bit of change to `networking` in the meantime but I tend to think the original argument still stands?

If we get cracking on this now it may be able to land in 3.6.

**Comment by robbmanes** [comment]
Hmm, I deleted that branch a while ago by mistake.  I do still think NetworkManager needs its own plugin - I can resubmit a branch in the next day or two after I take a look at the changes if that works for you.

**Comment by robbmanes** [comment]
Since I deleted my fork, I had to re-create the PR [here](https://github.com/sosreport/sos/pull/1255).  I'll close this one and we can continue to work in PR1255.
